src/engine/ExportQueryExecutionTrees.cpp

```diff
@@ -976,6 +974,153 @@ ad_utility::streams::stream_generator ExportQueryExecutionTrees::
+  for (auto& [string, index] : stringMapping_) {
+    sortedStrings[index] = string;
+  }
+  stringMapping_.clear();
+
+  std::string result;
+  // Rough estimate
+  result.reserve(sortedStrings.size() * 100);
+
+  for (const std::string& string : sortedStrings) {
+    absl::StrAppend(&result, raw(string.size()), string);
+  }
+
+  return result;
+}
+
+// _____________________________________________________________________________
+Id ExportQueryExecutionTrees::StringMapping::stringToId(
+    std::pair<std::string, const char*> optionalStringAndType) {
+  auto& [stringValue, xsdType] = optionalStringAndType;
+  if (xsdType != nullptr) {
+    absl::StrAppend(&stringValue, "^^", xsdType);
+  }
+  size_t distinctIndex = 0;
+  if (stringMapping_.contains(stringValue)) {
+    distinctIndex = stringMapping_.at(stringValue);
+  } else {
+    distinctIndex = stringMapping_[std::move(stringValue)] =
+        stringMapping_.size();
+  }
+  return Id::makeFromLocalVocabIndex(
+      reinterpret_cast<LocalVocabIndex>(distinctIndex));
+}
+
+// _____________________________________________________________________________
+Id ExportQueryExecutionTrees::toExportableId(Id originalId,
+                                             const QueryExecutionTree& qet,
+                                             const LocalVocab& localVocab,
+                                             StringMapping& stringMapping) {
+  if (isTrivial(originalId)) {
+    return originalId;
+  }
+  auto optionalStringAndType =
+      idToStringAndType(qet.getQec()->getIndex(), originalId, localVocab);
+  if (optionalStringAndType.has_value()) [[likely]] {
+    return stringMapping.stringToId(std::move(optionalStringAndType.value()));
+  }
+  return Id::makeUndefined();
+}
+
+// _____________________________________________________________________________
+template <>
+ad_utility::streams::stream_generator ExportQueryExecutionTrees::
+    selectQueryResultToStream<ad_utility::MediaType::binaryQleverExport>(
+        const QueryExecutionTree& qet,
+        const parsedQuery::SelectClause& selectClause,
+        LimitOffsetClause limitAndOffset,
+        CancellationHandle cancellationHandle) {
+  std::shared_ptr<const Result> result = qet.getResult(true);
+  result->logResultSize();
+  AD_LOG_DEBUG << "Starting binary export..." << std::endl;
+
+  // Magic bytes
+  co_yield "QLEVER.EXPORT";
+  // Export format version
+  co_yield raw(static_cast<uint16_t>(0));
+
+  // Get all columns with defined variables.
+  QueryExecutionTree::ColumnIndicesAndTypes columns =
+      qet.selectedVariablesToColumnIndices(selectClause, false);
+  std::erase(columns, std::nullopt);
+
+  // Export number of columns as a 16 bit unsigned int.
+  co_yield raw(static_cast<uint16_t>(columns.size()));
+
+  // Export actual variable names
+  for (const auto& optional : columns) {
+    const auto& variableName = optional.value().variable_;
+    co_yield raw(variableName.size());
+    co_yield variableName;
+  }
```

Comment:
are you aware of me having written a serialization frameword that can be used to get rid of all the reinterpret_cast(raw(blubb)) things?:)